Mathlib/Computability/QueryComplexity/Defs.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2025 Tomaz Mascarenhas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving, Tomaz Mascarenhas, Eric Wieser
+-/
+import Mathlib.Algebra.Group.Pi.Basic
+import Mathlib.Data.Set.Insert
+
+/-!
+# Definition of a model of computation based on oracles.
+
+This module defines an abstraction of computation with oracles, enabling the analysis of upper
+bounds on the query complexity of algorithms. It also provides a Monad instance for it.
+
+Note that this is the deterministic version.
+
+## Main Definitions
+
+* `QueryComplexity.Comp ι ω s α`: a computation with oracles in `s : Set ι` answering with `ω i`,
+  returning `α`.
+* `QueryComplexity.Comp.run f oracles : α × (I → ℕ)`: run `f : Comp ι s α` with
+  `oracles : I → Oracle ι`, returning the number of times each oracle is used.
+
+## Implementation notes
```

Comment:
I've recorded some of my hesitations with this design here. I think we can safely get all of the mergeSort PRs reviewed and merge, and then revisit these decisions.